catatonit: init at 0.1.5, fix podman --init

erosennin opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Podman uses [catatonit](https://github.com/openSUSE/catatonit) as the default container-init with `podman run --init`. It is currently broken on NixOS:

```
$ podman run --init busybox
Error: container-init binary not found on the host: stat /usr/libexec/podman/catatonit: no such file or directory
```

The idea is to add catatonit to nixpkgs, then patch podman to use `${catatonit}/bin/catatonit` instead of ` /usr/libexec/podman/catatonit`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

erosennin committed:
catatonit: init at 0.1.5

Signed-off-by: Andrey Golovizin <ag@sologoc.com>

erosennin commented:
I've added a check to ensure that the executable is statically linked, because it won't work inside a container otherwise. Using `readelf -d | grep` feels kind of sketchy though. Does anybody know a better way to do it?

r-rmcgibbo commented:
Result of `nixpkgs-review pr 114821` at a05a99c2 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/287ae8b6b37689e68d67bb49045d145e">catatonit</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`checkPhase` should probably contain `runHook preCheck` and `runHook postCheck`.

Near pkgs/applications/virtualization/catatonit/default.nix:27:3:
```
   |
27 |   checkPhase = ''
   |   ^
```
</li>
  </ul>
</details>

zowoq commented on pkgs/applications/virtualization/catatonit/default.nix:
```diff
@@ -0,0 +1,40 @@
+
+  src = fetchFromGitHub {
+    owner = "openSUSE";
+    repo = "catatonit";
```
```suggestion
    repo = pname;
```

erosennin replied:
Right, thanks.

zowoq commented on pkgs/applications/virtualization/catatonit/default.nix:
```diff
@@ -0,0 +1,40 @@
+    sha256 = "sha256-ciJ1MI7jr5P2PgxIykQ+BiwNUO8lQHGt0+U8CNbc5bI=";
+  };
+
+  patches = [
+    # Fix compilation with musl
+    (
+      fetchpatch {
+        url = "https://github.com/openSUSE/catatonit/commit/75014b1c3099245b7d0f44f24d7f6dc4888a45fd.patch";
+        sha256 = "sha256-9VMNUT1U90ocjvE7EXYfLxuodDwTXXHYg89qqa5Jq0g=";
+      }
+    )
+  ];
```
```suggestion
  patches = [
    # Fix compilation with musl
    (fetchpatch {
        url = "https://github.com/openSUSE/catatonit/commit/75014b1c3099245b7d0f44f24d7f6dc4888a45fd.patch";
        sha256 = "sha256-9VMNUT1U90ocjvE7EXYfLxuodDwTXXHYg89qqa5Jq0g=";
    })
  ];
```

zowoq commented on pkgs/applications/virtualization/catatonit/default.nix:
```diff
@@ -0,0 +1,40 @@
+  nativeBuildInputs = [ autoreconfHook ];
+  buildInputs = lib.optionals (!stdenv.hostPlatform.isMusl) [ glibc glibc.static ];
+
+  checkPhase = ''
+    runHook preCheck
+    readelf -d $out/bin/cacatonit | grep 'There is no dynamic section in this file.'
+    runHook postCheck
+  '';
```
```suggestion
  doInstallCheck = true;
  installCheckPhase = ''
    readelf -d $out/bin/cacatonit | grep 'There is no dynamic section in this file.'
  '';
```

zowoq commented on pkgs/applications/virtualization/catatonit/default.nix:
```diff
@@ -0,0 +1,40 @@
+  '';
+
+  meta = with lib; {
+    description = "A container init that is so simple it's effectively brain-dead.";
```
```suggestion
    description = "A container init that is so simple it's effectively brain-dead";
```

zowoq reviewed (commented):
cc @NixOS/podman

zowoq commented:
> patch podman ...

Shouldn't be needed, if adding it to the wrapper doesn't work it can be added to `containers.nix` as an explicit path with `init_path` like we do for CNI.

https://github.com/NixOS/nixpkgs/blob/80308388cd77ee58823c9b6f24b46892cd359145/nixos/modules/virtualisation/containers.nix#L119

erosennin commented:
@zowoq Thanks for the instant review!

saschagrunert reviewed (approved):
Yep yep, LGTM

adisbladis commented on pkgs/applications/virtualization/catatonit/default.nix:
```diff
@@ -0,0 +1,37 @@
+  ];
+
+  nativeBuildInputs = [ autoreconfHook ];
+  buildInputs = lib.optionals (!stdenv.hostPlatform.isMusl) [ glibc glibc.static ];
```
I don't think you need `glibc.static` explicitly.

erosennin replied:
Without `glibc.static` the executable would be linked to `/nix/store/.../libc.so.6` dynamically (checked it). We want a fully static binary so it can work inside any container.

adisbladis replied:
I find that a bit odd but I'll take your word for it. Thanks for humoring me :)

zowoq commented:
Adding it to the wrapper didn't work so it can be added here.
```diff
diff --git a/nixos/modules/virtualisation/containers.nix b/nixos/modules/virtualisation/containers.nix
index 997edf77ba9..4c7e365ac0f 100644
--- a/nixos/modules/virtualisation/containers.nix
+++ b/nixos/modules/virtualisation/containers.nix
@@ -118,8 +118,9 @@ in
       [network]
       cni_plugin_dirs = ["${pkgs.cni-plugins}/bin/"]
 
-      ${lib.optionalString (cfg.ociSeccompBpfHook.enable == true) ''
       [engine]
+      init_path = "${pkgs.catatonit}/bin/catatonit"
+      ${lib.optionalString (cfg.ociSeccompBpfHook.enable == true) ''
       hooks_dir = [
         "${config.boot.kernelPackages.oci-seccomp-bpf-hook}",
       ]
```

erosennin commented:
@zowoq thanks, I'll test it locally and try to add a NixOS test for `podman run --init`.

SuperSandro2000 commented on pkgs/applications/virtualization/catatonit/default.nix:
```diff
@@ -0,0 +1,37 @@
+{ lib, stdenv, fetchFromGitHub, fetchpatch, autoreconfHook, glibc }:
+
+stdenv.mkDerivation rec {
+  version = "0.1.5";
+  pname = "catatonit";
```
```suggestion
  pname = "catatonit";
  version = "0.1.5";
```

SuperSandro2000 commented on pkgs/applications/virtualization/catatonit/default.nix:
```diff
@@ -0,0 +1,37 @@
+    owner = "openSUSE";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-ciJ1MI7jr5P2PgxIykQ+BiwNUO8lQHGt0+U8CNbc5bI=";
```
```suggestion
    sha256 = "ciJ1MI7jr5P2PgxIykQ+BiwNUO8lQHGt0+U8CNbc5bI=";
```

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -192,6 +192,8 @@ in
 
   castxml = callPackage ../development/tools/castxml { };
 
+  catatonit = callPackage ../applications/virtualization/catatonit {};
```
```suggestion
  catatonit = callPackage ../applications/virtualization/catatonit { };
```

SuperSandro2000 commented:
``nix-build -A pkgsMusl.catatonit`` builds fine.

erosennin commented:
Is anything required from my side?

zowoq committed:
nixos/containers: add catatonit / init_path

https://github.com/containers/common/blob/master/docs/containers.conf.5.md

- Also drop unneeded true from ociSeccompBpfHook

adisbladis commented:
@GrahamcOfBorg test podman

erosennin committed:
nixos/tests/podman: test podman run --init

erosennin commented:
Added a test for `podman run --init` while we are at it. I actually thoght that `podman --init` stuff should go into a separate PR but why not. :)

adisbladis commented:
@GrahamcOfBorg test podman

zowoq commented:
Probably want to add passthru.tests to catatonic and also add it to the podman team as it's included in the module by default. I can do both in a few hours.

zowoq committed:
catatonit: add `passthru.tests` and `teams.podman`

adisbladis commented on pkgs/applications/virtualization/catatonit/default.nix:
```diff
@@ -27,11 +27,13 @@ stdenv.mkDerivation rec {
     readelf -d $out/bin/catatonit | grep 'There is no dynamic section in this file.'
   '';
 
+  passthru.tests = { inherit (nixosTests) podman; };
```
That's fine I think. We can always add more tests to explicitly test `catatonit`. In the mean time it's good if ofborg knows to run podman tests whenever catatonit is bumped.
